File: epi_judge_python/power_x_y.py
from test_framework import generic_test


def power(x: float, y: int) -> float:
    if y < 0:
        x, y = 1 / x, -y
    if y == 0:
        return 1
    if y & 1:
        return x * power(x, y - 1)
    else:
        half_power = power(x, y >> 1)
        return half_power * half_power


if __name__ == '__main__':
    exit(generic_test.generic_test_main('power_x_y.py', 'power_x_y.tsv',
                                        power))

File: epi_judge_cpp/power_x_y.cc
#include "test_framework/generic_test.h"

double Power(double x, int y) {
    if (y < 0) {
        x = 1 / x;
        y = -y;
    }
    double ans = 1;
    double mult = x;
    while (y) {
        if (y & 1) {
            ans *= mult;
        }
        mult *= mult;
        y /= 2;
    }
    return ans;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"x", "y"};
    return GenericTestMain(args, "power_x_y.cc", "power_x_y.tsv", &Power,
                           DefaultComparator{}, param_names);
}
